test/upgrade/testing_operations_test.go

```diff
@@ -0,0 +1,308 @@
+}
+
+func generalizeOpsFromBg(ops []upgrade.BackgroundOperation) operations {
+	gen := make([]*operation, len(ops))
+	for idx, op := range ops {
+		gen[idx] = &operation{bg: op}
+	}
+	return operations{ops: gen}
+}
+
+func createMessages(mf formats) messages {
+	return messages{
+		skipped: func(args ...interface{}) string {
+			empty := ""
+			if mf.skipped == empty {
+				return empty
+			}
+			return fmt.Sprintf(mf.skipped, args...)
+		},
+		starting: func(args ...interface{}) string {
+			return fmt.Sprintf(mf.starting, args...)
+		},
+		element: func(args ...interface{}) string {
+			return fmt.Sprintf(mf.element, args...)
+		},
+	}
+}
+
+func (tt *texts) append(msgs ...string) *texts {
+	for _, msg := range msgs {
+		if msg == "" {
+			continue
+		}
+		tt.elms = append(tt.elms, msg)
+	}
+	return tt
+}
+
+func completeSuiteExample(fp failurePoint) upgrade.Suite {
+	suite := upgrade.Suite{
+		Tests: upgrade.Tests{
+			PreUpgrade: []upgrade.Operation{
+				serving.tests.preUpgrade, eventing.tests.preUpgrade,
+			},
+			PostUpgrade: []upgrade.Operation{
+				serving.tests.postUpgrade, eventing.tests.postUpgrade,
+			},
+			PostDowngrade: []upgrade.Operation{
+				serving.tests.postDowngrade, eventing.tests.postDowngrade,
+			},
+			ContinualTests: []upgrade.BackgroundOperation{
+				serving.tests.continual, eventing.tests.continual,
+			},
+		},
+		Installations: upgrade.Installations{
+			Base: []upgrade.Operation{
+				serving.installs.stable, eventing.installs.stable,
+			},
+			UpgradeWith: []upgrade.Operation{
+				serving.installs.head, eventing.installs.head,
+			},
+			DowngradeWith: []upgrade.Operation{
+				serving.installs.stable, eventing.installs.stable,
+			},
+		},
+	}
+	return enrichSuiteWithFailures(suite, fp)
+}
+
+func emptySuiteExample() upgrade.Suite {
+	return upgrade.Suite{
+		Tests:         upgrade.Tests{},
+		Installations: upgrade.Installations{},
+	}
+}
+
+func enrichSuiteWithFailures(suite upgrade.Suite, fp failurePoint) upgrade.Suite {
+	steps := createSteps(suite)
+	for i, st := range steps {
+		for j, op := range st.ops.ops {
+			if fp.step == i+1 && fp.element == j+1 {
+				op.fail(fp.step == 3)
+			}
+		}
+	}
+	return recreateSuite(steps)
+}
+
+func recreateSuite(steps []*step) upgrade.Suite {
+	suite := &upgrade.Suite{
+		Tests:         upgrade.Tests{},
+		Installations: upgrade.Installations{},
+	}
+	for _, st := range steps {
+		st.updateSuite(st.ops, suite)
+	}
+	return *suite
+}
+
+func (o operation) Name() string {
+	if o.op != nil {
+		return o.op.Name()
+	} else {
+		return o.bg.Name()
+	}
+}
+
+func (o *operation) fail(setupFail bool) {
+	failureTestingMessage := "This error is expected to be seen. Upgrade suite should fail."
+	testName := fmt.Sprintf("FailingOf%s", o.Name())
+	if o.op != nil {
+		prev := o.op
+		o.op = upgrade.NewOperation(testName, func(c upgrade.Context) {
+			handler := prev.Handler()
+			handler(c)
+			c.T.Error(failureTestingMessage)
+			c.Log.Error(failureTestingMessage)
```

Comment:
I'm curious why we use this pattern of logging the error in both `T` and `Log`. Is the user supposed to do it the same way in the end? Would that make sense to do it only once then then propagate the error automatically? Just for my understanding

Reply:
Oh. This is just for the verbosity of what's going on in those tests. Especially for failure testing. I capture T stdout and Log stdout separately, and use it to validate where given output was received. Log stdout is also printed for easier debugging. 

That's why it is in both places. In real tests you should use only `t.Error`.